askup/tests/test_functional/test_general.py

```diff
@@ -1443,3 +1459,169 @@ def test_user_has_no_questions(self):
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'unsubscribe'}
+            )
+        )
+
+        # Unsubscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'unsubscribe').json()
+        self.assertEqual(response['result'], 'success')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'subscribe'}
+            )
+        )
+
+    @client_user('student01', 'student01')
+    def test_subscription_fail(self):
+        """
+        Test failed on permission basis subscription.
+        """
+        qset_id = 10  # Qset 4-1 from the mockups
+        # Subscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'subscribe').json()
+        self.assertEqual(response['result'], 'fail')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'subscribe'}
+            )
+        )
+
+        # Unsubscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'unsubscribe').json()
+        self.assertEqual(response['result'], 'fail')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'unsubscribe'}
+            )
+        )
+
+
+class TestSubscriptionsMailing(TestCase):
+    """
+    Testing the subject subscription mailing functionality.
+    """
+
+    fixtures = ['groups', 'mockup_data']
+
+    def setUp(self):
+        """
+        Set up the test conditions.
+        """
+        settings.DEBUG = False
+
+    @client_user('student01', 'student01')
+    @patch('askup.utils.general.do_send_subscriptions_to_recipients')
+    def test_subscriptions_sending_no_actual_subscriptions(self, mock_do_send_subscriptions_to_recipients):
+        """
+        Test subscriptions sending.
+        """
+        send_subscription_emails()
+        mock_do_send_subscriptions_to_recipients.return_value = None
+        mock_do_send_subscriptions_to_recipients.assert_not_called()
+
+    @client_user('student01', 'student01')
+    @patch('askup.utils.general.do_send_subscriptions_to_recipients')
+    def test_subscriptions_sending_with_subscriptions(self, mock_do_send_subscriptions_to_recipients):
+        """
+        Test subscriptions sending.
+        """
+        qset_id_1 = 4  # Qset 1-1
+        qset_id_2 = 5  # Qset 1-2
+        text_1 = 'Test question 1-1'
+        text_2 = 'Test question 1-2'
+        answer_text_1 = 'Test question answer 1-1'
+        answer_text_2 = 'Test question answer 1-2'
+        QuestionModelFormTest.qset_create_question(self, text_1, answer_text_1, qset_id_1)
```

Comment:
This is an interesting pattern, by the way, sending in `self` (referring to the `TestSubscriptionsMailing` test case) as a proxy for the `QuestionModelFormTest`'s `self` reference, and using `staticmethod`s in the `QuestionModelFormTest`.